KEYCLOAK-14988 - Add generate-action-token service endpoint

waigel opened the pull request:
generate  tokens via rest api to perform actions

pedroigor commented on server-spi/src/main/java/org/keycloak/models/ActionToken.java:
```diff
@@ -0,0 +1,43 @@
+
+package org.keycloak.models;
+
+public class ActionToken {
```
I would just avoid having this class but building the JSON response with the "token" claim.

VerHext replied:
You're right. I have now removed the class.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -845,6 +846,56 @@ public Response executeActionsEmail(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_P
+    @Path("generate-actions-token")
+    @PUT
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response executeActionsEmail(@QueryParam(OIDCLoginProtocol.CLIENT_ID_PARAM) String clientId,
```
Instead of replicating code from `org.keycloak.services.resources.admin.UserResource#executeActionsEmail` we could probably reuse the same code that performs the validation and token creation.

I would also change the method name to `createActionToken` or something.

VerHext replied:
Outsourcing of validation and token generation to `createActionToken` function.

pedroigor commented:
@VerHext Thanks. I see the reasoning behind it and I can't think about any negative impact (mainly in terms of security) that this could bring.

But I would like to listen from others.

fidgi commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -753,32 +753,28 @@ public Response resetPasswordEmail(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PA
+     * Create action token 
+     * 
+     * 
+     * An token to perform a set of required actions. The clientId parameters are
```
Suggested rewording : "**A** token ... . The clientId parameter **is** optional"

fidgi commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -845,6 +870,28 @@ public Response executeActionsEmail(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_P
+     * Returns a token to perform a set of required actions. The clientId parameters are
+     * optional.
+     *
+     * @param clientId Client id
```
missing redirectUri parameter in the JavaDoc

fidgi commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -845,6 +870,28 @@ public Response executeActionsEmail(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_P
+     /**
+     * Generate action tokens endpoint
+     *
+     * Returns a token to perform a set of required actions. The clientId parameters are
```
Just replicate the suggested JavaDoc from createActionToken method

fidgi commented:
@VerHext @pedroigor This PR is comming just in time. We're looking for this functionnality for our project. It'd be great if it can be added to the next release.

pedroigor reviewed (approved):
@stianst I'm approving. Looks like a simple addition that opens the doors for use cases where notifications are sent outside Keycloak.

fidgi commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -845,6 +869,29 @@ public Response executeActionsEmail(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_P
+     * @return
+     */
+    @Path("generate-action-token")
+    @PUT
```
Why PUT rather than POST for the creation of the action token ? If each call generates a new token, POST may be more appropriate.

VerHext replied:
I expected that the function replaced the action token. Then it would be in my opinion PUT the method.
But we can rename them if POST is the better method.

pedroigor replied:
Good point from @fidgi. Could you change this?

VerHext replied:
Changed

pedroigor reviewed (changes requested):
Other than the HTTP method change, LGTM.

waigel committed:
- rename function "executeActionsEmail" to "generateActionToken"
- outsource validation and token creation to "createActionToken" function
- remove ActionToken class
- fix javadoc
- change http method to POST
- add to admin-client

stianst reviewed (changes requested):
Action tokens should be sent to the user, not the admin.

Looking at the linked in JIRA and Discourse question the use-case here is to create a custom verify email "email", which is already supported. You can simply create a custom email theme, with a custom message, or you can create your own template if you'd like. There's no need to have something outside of KC generate the token, and send the email.

VerHext commented:
For example, we create accounts outside of Keycloak and would like to include an action token in the welcome email. In this use case it is easier to generate the token directly. In addition, we would also send the token for resetting the password via a different channel instead of e-mails. To ensure more security.

stianst commented:
@VerHext You can already trigger sending the action token to the user after creating the user out of bounds. I can see the case of sending the action token over something other than email, but that should be done in a different way to this. I'd be very open to support KC communicating with the user over different mechanisms than email. What mechanisms do you use specifically?

stianst commented:
@fidgi you can already achieve that by just sending the action token email to the user, so you don't need to generate and return an action token to the admin.

stianst reviewed (changes requested):
I'm against allowing returning an action token to the admin directly. As that would allow an admin to perform the action on behalf of the user, which there is no guarantees that the admin should be allowed to do.

fidgi commented:
@stianst Can you please elaborate on how to send an email with an action token from outside keycloak ? In our case, an external web app should be able to send the action token by any means that fit its needs (e.g. mail , SMS ) .The user is not authenticated since he doesn't have completed the sign-in process yet.

stianst commented:
Right now you are limited to email only to communicate with the user. That can somewhat be workaround by having a custom EmailSenderProvider, that would use SMS or email internally, but that is for sure a bit of a workaround. We would like to have support for SMS built-in to KC as well.

There's an endpoint that sends an action token email to the user. Take a look at docs, or the admin console (with browsers developer tools) to figure out how this is done.

fidgi commented:
@stianst understood . May be I should look at the Application Initiated Action to solve my use cases.
